src/ddmd/traits.d

```diff
@@ -403,6 +403,33 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
         return new IntegerExp(e.loc, false, Type.tbool);
     }
 
+    TypeFunction firstArgFunctionType()
```

Comment:
IMO a better abstraction is `toTypeFunction(RootObject o)` and passing the validated parameter into this function. That would allow it not being a nested function and used elsewhere, too. 

Even better, I'd split it into two overloads, one for Type, and one for Dsymbol, as you now have widen the allowed parameters to traits that support types only so far.